R/selections.R

```diff
@@ -106,27 +107,30 @@ eval_select_recipes <- function(quos, data, info) {
 
   expr <- expr(c(!!!quos))
 
-  # FIXME: Ideally this is `FALSE`, but empty selections incorrectly throw an
+  # FIXME: Ideally this is `FALSE` for strict selection,
+  # but empty selections incorrectly throw an
   # error when this is false due to the following bug:
   # https://github.com/r-lib/tidyselect/issues/221
-  allow_rename <- TRUE
+  # Once it's fixed, remove this and pass allow_rename to
+  # tidyselect::eval_select().
+  allow_rename_ts <- TRUE
 
   sel <- tidyselect::eval_select(
     expr = expr,
     data = data,
-    allow_rename = allow_rename
+    allow_rename = allow_rename_ts
   )
 
   # Return names not positions, as these names are
   # used for both the training and test set and their positions
-  # may have changed. `sel` won't be named because when `allow_rename = FALSE`,
-  # `eval_select()` returns an unnamed vector.
+  # may have changed. If renaming is allowed, add the new names.
   out <- names(data)[sel]
+  if (allow_rename) names(out) <- names(sel)
```

Comment:
To be consistent with tidyselect, I think that we should _always_ add `names(sel)` onto `names(out)`. I realize that in the `allow_rename = FALSE` case, this is a bit redundant, but I like that it is all around more consistent.

```r
library(tidyselect)

eval_select(quote(mpg), mtcars)
#> mpg 
#>   1
eval_select(quote(c(new = mpg)), mtcars)
#> new 
#>   1

# for us i think it would look like:

# eval_select_recipes(quote(mpg))
# mpg
# "mpg"

# eval_select_recipes(quote(c(new = mpg)), allow_rename = TRUE)
# new
# "mpg"
```

That will also make it cleaner when the tidyselect bugs finally do get fixed. At that point, the only place that `allow_rename` will be used is where we pass it on to `eval_select()`.

Reply:
Sounds good to me. If we do that, several tests fail though so I'd open a new issue for that and address that in a separate PR.

Reply:
Yea, as discussed offline, we will come back to this elsewhere if we need to!

Reply:
on a second thought: if consistency with tidyselect is the aim, then maybe always returning a named vector is not what we want after all, given that `eval_select()` returns an unnamed vector when `allow_rename = FALSE`.

``` r
library(tidyselect)

eval_select(quote(mpg), mtcars, allow_rename = FALSE)
#> [1] 1
```

<sup>Created on 2021-03-03 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

Reply:
There is a decent chance that is a bug. I ran across that awhile back https://github.com/r-lib/tidyselect/issues/220

Reply:
Ah, I see :+1:  I've turned your comment into an issue:  #654